input/otlp/metadata.go

```diff
@@ -545,3 +547,21 @@ func setLabel(key string, event *modelpb.APMEvent, v pcommon.Value) {
 		}
 	}
 }
+
+// Sanitize the datastream fields (dataset, namespace) to apply restrictions
+// as outlined in https://www.elastic.co/guide/en/ecs/current/ecs-data_stream.html
+func sanitizeDataStreamField(field string) string {
+	field = strings.ToLower(field)
+	field = strings.Map(replaceReservedRune, field)
+	if len(field) > MaxDataStreamBytes {
+		return field[:MaxDataStreamBytes]
+	}
+	return field
+}
+
+func replaceReservedRune(r rune) rune {
+	if strings.ContainsRune(DisallowedDataStreamRunes, r) {
+		return '_'
+	}
+	return r
```

Comment:
personally i'd switch case a rune (or byte behind a multi-byte rune check), hoping that compiler is smart enough to create a jump table. Otherwise, @1pkg 's ascii table map trick is as good.

> Having a list of disallowed runes in this switch case, as well as in DisallowDataStreamRunes creates 2 sources of truth and will be harder to maintain down the road, especially when the 2 lines are separated by 200+ lines.

Revisiting my earlier comment, @rubvs actually you have a solution very close to what I'd do initially. My duplication concern comes from 2 sources of truth in actual code file. But if the 2 sources are spread across actual code and test code, I'd be less concerned.

Reply:
> personally i'd switch case a rune (or byte behind a multi-byte rune check), hoping that compiler is smart enough to create a jump table

I think the issue here is optimizing the comparison. A rangetable would create efficient comparison and should perform quite well if the number of runes to compare is high enough, however, since we have only a few here the overheads would probably drown any gains or make it comparable. We are probably talking about a few nanosecs here so not much of a difference anyway. Not sure how the switch-case will look like though since it would technically be all the runes in one case.

Reply:
> Not sure how the switch-case will look like though since it would technically be all the runes in one case.

That's how I imagine.

+1 on performance difference here. I'm fine with the ContainsRune here since my bench also shows ns-level baseline